torchrl/modules/planners/cem.py

```diff
@@ -0,0 +1,133 @@
+    The CEM planning step is repeated for a specified number of steps.
+    At the end, we recover the best action which is the one that maximizes the reward given a planning horizon.
+
+    Args:
+        env (Environment): The environment to perform the planning step on (Can be ModelBasedEnv or EnvBase).
+        planning_horizon (int): The number of steps to perform the planning step for.
+        optim_steps (int): The number of steps to perform the MPC planning step for.
+        num_candidates (int): The number of candidates to sample from the Gaussian distribution.
+        num_top_k_candidates (int): The number of top candidates to use to update the mean and standard deviation of the Gaussian distribution.
+        reward_key (str): The key in the TensorDict to use to retrieve the reward.
+        action_key (str): The key in the TensorDict to use to store the action.
+
+    Returns:
+        TensorDict: The TensorDict with the action added.
+    """
+
+    def __init__(
+        self,
+        env: EnvBase,
+        planning_horizon: int,
+        optim_steps: int,
+        num_candidates: int,
+        num_top_k_candidates: int,
+        reward_key: str = "reward",
+        action_key: str = "action",
+    ):
+        super().__init__(env=env, action_key=action_key)
+        self.planning_horizon = planning_horizon
+        self.optim_steps = optim_steps
+        self.num_candidates = num_candidates
+        self.num_top_k_candidates = num_top_k_candidates
+        self.reward_key = reward_key
+
+    def planning(self, td: TensorDictBase) -> torch.Tensor:
+        batch_size = td.batch_size
+        expanded_original_td = (
+            td.expand(*batch_size, self.num_candidates).view(-1)
+        )
+        flatten_batch_size = batch_size.numel()
+        actions_means = torch.zeros(
+            flatten_batch_size,
+            1,
+            self.planning_horizon,
+            *self.action_spec.shape,
+            device=td.device,
+            dtype=self.env.action_spec.dtype,
+        )
+        actions_stds = torch.ones(
+            flatten_batch_size,
+            1,
+            self.planning_horizon,
+            *self.action_spec.shape,
+            device=td.device,
+            dtype=self.env.action_spec.dtype,
+        )
+        for _ in range(self.optim_steps):
+            actions = actions_means + actions_stds * torch.randn(
+                flatten_batch_size,
+                self.num_candidates,
+                self.planning_horizon,
+                *self.action_spec.shape,
+                device=td.device,
+                dtype=self.env.action_spec.dtype,
+            )
+            actions = actions.view(
+                flatten_batch_size * self.num_candidates,
+                self.planning_horizon,
+                *self.action_spec.shape,
+            )
+            actions = self.env.action_spec.project(actions)
+            optim_td = expanded_original_td.clone()
```

Comment:
I guess that here you want to_tensordict instead?
clone for tensordicts doesn't work the way it does for torch.Tensor (it doesn't always materialize the lazy operations etc)